src/Razor/test/Microsoft.AspNetCore.Razor.LanguageServer.Test/Diagnostics/RazorPullDiagnosticsEndpointTest.cs

```diff
@@ -0,0 +1,169 @@
+                " @{ void Foo< @* Comment! *@ TValue>() {} }  ",
+                "    void Foo<  TValue>() {} ",
+                new[] {
+                    // "Foo< "
+                    new SourceMapping(
+                        new SourceSpan(3, 11),
+                        new SourceSpan(3, 11)),
+
+                    // " TValue>"
+                    new SourceMapping(
+                        new SourceSpan(28, 14),
+                        new SourceSpan(15, 13)),
+                });
+            var documentContext = CreateDocumentContext(uri, codeDocument);
+            var requestContext = CreateRazorRequestContext(documentContext);
+
+            // Act
+            var result = await endpoint.HandleRequestAsync(request, requestContext, DisposalToken);
+
+            // Assert
+            Assert.Null(result);
+        }
+
+        [Fact]
+        public async Task Handle_CSharpDiagnostics_Remapped()
+        {
+            // Arrange
+            var uri = new Uri("C:/path/to/document.cshtml");
+            var cSharpDiagnostics = new VSInternalDiagnosticReport[]
+            {
+                new VSInternalDiagnosticReport{
+                    Diagnostics = new Diagnostic[]{
+                        new Diagnostic
+                        {
+                            Message = "Message",
+                            Range = new Range
+                            {
+                                Start = new Position(0, 15),
+                                End = new Position(0, 21),
+                            },
+                        },
+                    },
+                },
+            };
+            var languageServer = new Mock<ClientNotifierServiceBase>(MockBehavior.Strict);
+            languageServer
+                .Setup(client => client.SendRequestAsync<IDelegatedParams, IEnumerable<VSInternalDiagnosticReport>?>(
+                    RazorLanguageServerCustomMessageTargets.RazorPullDiagnosticEndpointName, It.IsAny<DelegatedDiagnosticParams>(), DisposalToken))
+                .ReturnsAsync(cSharpDiagnostics);
+            var endpoint = GetEndpoint(useSingleServer: true, languageServer.Object);
+            var request = new VSInternalDocumentDiagnosticsParamsBridge
+            {
+                TextDocument = new TextDocumentIdentifier
+                {
+                    Uri = uri,
+                },
+            };
+            var codeDocument = CreateCodeDocumentWithCSharpProjection(
+                " @{ void Foo< @* Comment! *@ TValue>() {} }  ",
+                "    void Foo<  TValue>() {} ",
+                new[] {
+                    // "Foo< "
+                    new SourceMapping(
+                        new SourceSpan(3, 11),
+                        new SourceSpan(3, 11)),
+
+                    // " TValue>"
+                    new SourceMapping(
+                        new SourceSpan(28, 14),
+                        new SourceSpan(15, 13)),
+                });
```

Comment:
Doesn't have to be in this PR, but you should check out `TestFileMarkupParser`. Makes this sort of thing much easier. eg:
https://github.com/dotnet/razor-tooling/blob/fef5e7ff27f70db71b311dd8cb056c29c419d7e7/src/Razor/test/Microsoft.AspNetCore.Razor.LanguageServer.Test/Definition/RazorDefinitionEndpointTest.cs#L318